File: Tests/iOS Tests/PNPublishWithHistoryTests.m
//
//  PNPublishWithHistoryTests.m
//  PubNub Tests
//
//  Created by Vadim Osovets on 6/19/15.
//
//

#import <UIKit/UIKit.h>
#import <XCTest/XCTest.h>
#import <PubNub/PubNub.h>

#import "PNBasicClientTestCase.h"

#import "NSString+PNTest.h"

@interface PNPublishWithHistoryTests : PNBasicClientTestCase

@end

@implementation PNPublishWithHistoryTests

- (BOOL)isRecording{
    return NO;
    
}

- (NSString *)publishChannelString {
    return @"9BA810C6-985D-4797-926F-CC81749CC774";
}

- (void)testSimplePublishWithHistory {
    [self performVerifiedPublish:@"test" onChannel:[self publishChannelString]
                  storeInHistory:YES
                  withAssertions:^(PNPublishStatus *status) {
                      XCTAssertNotNil(status);
                      XCTAssertEqual(status.category, PNAcknowledgmentCategory);
                      XCTAssertEqual(status.operation, PNPublishOperation);
                      XCTAssertEqual(status.statusCode, 200);
                      XCTAssertFalse(status.isError);
                      NSLog(@"status.data.information: %@", status.data.information);
                      NSLog(@"status.data.timeToken: %@", status.data.timetoken);
                      XCTAssertEqualObjects(status.data.information, @"Sent");
                      XCTAssertEqualObjects(status.data.timetoken, @"14355338601176506");
    }];
}

- (void)testSimplePublishWithoutHistory {
    [self performVerifiedPublish:@"test" onChannel:[self publishChannelString]
                  storeInHistory:NO
                  withAssertions:^(PNPublishStatus *status) {
                      XCTAssertNotNil(status);
                      XCTAssertEqual(status.category, PNAcknowledgmentCategory);
                      XCTAssertEqual(status.operation, PNPublishOperation);
                      XCTAssertEqual(status.statusCode, 200);
                      XCTAssertFalse(status.isError);
                      NSLog(@"status.data.information: %@", status.data.information);
                      NSLog(@"status.data.timeToken: %@", status.data.timetoken);
                      XCTAssertEqualObjects(status.data.information, @"Sent");
                      XCTAssertEqualObjects(status.data.timetoken, @"14355338601978693");
                  }];
}

- (void)testPublishWithHistoryNilMessage {
    [self performVerifiedPublish:nil
                       onChannel:[self publishChannelString]
                  storeInHistory:YES
                  withAssertions:^(PNPublishStatus *status) {
        XCTAssertNotNil(status);
        XCTAssertEqual(status.category, PNBadRequestCategory);
        XCTAssertEqual(status.operation, PNPublishOperation);
        XCTAssertEqual(status.statusCode, 400);
        XCTAssertTrue(status.isError);
        NSLog(@"status.data.information: %@", status.data.information);
        NSLog(@"status.data.timeToken: %@", status.data.timetoken);
        XCTAssertNil(status.data.information);
        XCTAssertNil(status.data.timetoken);
    }];
}

- (void)testPublishWithoutHistoryNilMessage {
    [self performVerifiedPublish:nil
                       onChannel:[self publishChannelString]
                  storeInHistory:NO
                  withAssertions:^(PNPublishStatus *status) {
                      XCTAssertNotNil(status);
                      XCTAssertEqual(status.category, PNBadRequestCategory);
                      XCTAssertEqual(status.operation, PNPublishOperation);
                      XCTAssertEqual(status.statusCode, 400);
                      XCTAssertTrue(status.isError);
                      NSLog(@"status.data.information: %@", status.data.information);
                      NSLog(@"status.data.timeToken: %@", status.data.timetoken);
                      XCTAssertNil(status.data.information);
                      XCTAssertNil(status.data.timetoken);
                  }];
}

- (void)testPublishDictionaryWithHistory {
    [self performVerifiedPublish:@{@"test" : @"test"}
                       onChannel:[self publishChannelString]
                  storeInHistory:YES
                  withAssertions:^(PNPublishStatus *status) {
                      XCTAssertNotNil(status);
                      XCTAssertEqual(status.category, PNAcknowledgmentCategory);
                      XCTAssertEqual(status.operation, PNPublishOperation);
                      XCTAssertEqual(status.statusCode, 200);
                      XCTAssertFalse(status.isError);
                      NSLog(@"status.data.information: %@", status.data.information);
                      NSLog(@"status.data.timeToken: %@", status.data.timetoken);
                      XCTAssertEqualObjects(status.data.information, @"Sent");
                      XCTAssertEqualObjects(status.data.timetoken, @"14355338598726743");
    }];
}

- (void)testPublishDictionaryWithoutHistory {
    [self performVerifiedPublish:@{@"test" : @"test"}
                       onChannel:[self publishChannelString]
                  storeInHistory:NO
                  withAssertions:^(PNPublishStatus *status) {
                      XCTAssertNotNil(status);
                      XCTAssertEqual(status.category, PNAcknowledgmentCategory);
                      XCTAssertEqual(status.operation, PNPublishOperation);
                      XCTAssertEqual(status.statusCode, 200);
                      XCTAssertFalse(status.isError);
                      NSLog(@"status.data.information: %@", status.data.information);
                      NSLog(@"status.data.timeToken: %@", status.data.timetoken);
                      XCTAssertEqualObjects(status.data.information, @"Sent");
                      XCTAssertEqualObjects(status.data.timetoken, @"14355338599684259");
                  }];
}

- (void)testPublishToNilChannelWithHistory {
    [self performVerifiedPublish:@{@"test" : @"test"}
                       onChannel:nil
                  storeInHistory:YES
                  withAssertions:^(PNPublishStatus *status) {
        XCTAssertNotNil(status);
        XCTAssertEqual(status.category, PNBadRequestCategory);
        XCTAssertEqual(status.operation, PNPublishOperation);
        XCTAssertEqual(status.statusCode, 400);
        XCTAssertTrue(status.isError);
        NSLog(@"status.data.information: %@", status.data.information);
        NSLog(@"status.data.timeToken: %@", status.data.timetoken);
        XCTAssertNil(status.data.information);
        XCTAssertNil(status.data.timetoken);
    }];
}

- (void)testPublishToNilChannelWithoutHistory {
    [self performVerifiedPublish:@{@"test" : @"test"}
                       onChannel:nil
                  storeInHistory:NO
                  withAssertions:^(PNPublishStatus *status) {
                      XCTAssertNotNil(status);
                      XCTAssertEqual(status.category, PNBadRequestCategory);
                      XCTAssertEqual(status.operation, PNPublishOperation);
                      XCTAssertEqual(status.statusCode, 400);
                      XCTAssertTrue(status.isError);
                      NSLog(@"status.data.information: %@", status.data.information);
                      NSLog(@"status.data.timeToken: %@", status.data.timetoken);
                      XCTAssertNil(status.data.information);
                      XCTAssertNil(status.data.timetoken);
                  }];
}

#pragma mark - Main flow

- (void)performVerifiedPublish:(id)message
                     onChannel:(NSString *)channel
                storeInHistory:(BOOL)storeInHistory
                withAssertions:(PNPublishCompletionBlock)verificationBlock {
    XCTestExpectation *networkExpectation = [self expectationWithDescription:@"network"];
    [self.client publish:message toChannel:channel storeInHistory:storeInHistory withCompletion:^(PNPublishStatus *status) {
        verificationBlock(status);
        [networkExpectation fulfill];
    }];
    [self waitForExpectationsWithTimeout:5 handler:^(NSError *error) {
        if (error) {
            NSLog(@"error: %@", error);
            XCTFail(@"what went wrong?");
        }
    }];
}

@end

File: Tests/iOS Tests-Swift/PNPublishWithHistoryTests.swift
//
//  PNPublishWithHistoryTests.swift
//  PubNub Tests
//
//  Created by Vadim Osovets on 09/22/15.
//
//

import UIKit
import XCTest

class PNPublishWithHistoryTests: PNBasicClientTestCase {
    
    
    let publishChannelString = "9BA810C6-985D-4797-926F-CC81749CC774"
        
    override func isRecording() -> Bool {
        return false
    }
    
    func testSimplePublishWithHistory() {
        self.performVerifiedPublish("test", onChannel: publishChannelString, storeInHistory: true) { (status) -> Void in
            
            XCTAssertNotNil(status)
            XCTAssertEqual(status.category, PNStatusCategory.PNAcknowledgmentCategory)
            XCTAssertEqual(status.operation, PNOperationType.PublishOperation)
            XCTAssertEqual(status.statusCode, 200)
            XCTAssertFalse(status.error)
            XCTAssertEqual(status.data.information, "Sent")
            XCTAssertEqual(status.data.timetoken, NSDecimalNumber(string: "14355338601176506"))
        }
    }
    
    func testSimplePublishWithoutHistory() {
        self.performVerifiedPublish("test", onChannel: publishChannelString, storeInHistory: false) { (status) -> Void in
            
            XCTAssertNotNil(status)
            XCTAssertEqual(status.category, PNStatusCategory.PNAcknowledgmentCategory)
            XCTAssertEqual(status.operation, PNOperationType.PublishOperation)
            XCTAssertEqual(status.statusCode, 200)
            XCTAssertFalse(status.error)
            XCTAssertEqual(status.data.information, "Sent")
            XCTAssertEqual(status.data.timetoken, NSDecimalNumber(string: "14355338601978693"))
        }
    }
    
    func testPublishWithoutHistoryNilMessage() {
        self.performVerifiedPublish(nil, onChannel: publishChannelString, storeInHistory: false) { (status) -> Void in
            
            XCTAssertNotNil(status)
            XCTAssertTrue(status.category == .PNBadRequestCategory)
            XCTAssertTrue(status.operation == .PublishOperation)
            XCTAssertEqual(status.statusCode, 400)
            XCTAssertTrue(status.error)
            XCTAssertNil(status.data)
        }
    }
    
    func testPublishDictionaryWithHistory() {
        self.performVerifiedPublish(["test" : "test"], onChannel: publishChannelString, storeInHistory: true) { (status) -> Void in
            
            XCTAssertNotNil(status)
            XCTAssertEqual(status.category, PNStatusCategory.PNAcknowledgmentCategory)
            XCTAssertEqual(status.operation, PNOperationType.PublishOperation)
            XCTAssertEqual(status.statusCode, 200)
            XCTAssertFalse(status.error)
            XCTAssertEqual(status.data.information, "Sent")
            XCTAssertEqual(status.data.timetoken, NSDecimalNumber(string: "14355338598726743"))
        }
    }
    
    func testPublishDictionaryWithoutHistory() {
        self.performVerifiedPublish(["test" : "test"], onChannel: publishChannelString, storeInHistory: false) { (status) -> Void in
            
            XCTAssertNotNil(status)
            XCTAssertEqual(status.category, PNStatusCategory.PNAcknowledgmentCategory)
            XCTAssertEqual(status.operation, PNOperationType.PublishOperation)
            XCTAssertEqual(status.statusCode, 200)
            XCTAssertFalse(status.error)
            XCTAssertEqual(status.data.information, "Sent")
            XCTAssertEqual(status.data.timetoken, NSDecimalNumber(string: "14355338599684259"))
        }
    }
    
    func testPublishToNilChannelWithHistory() {
        self.performVerifiedPublish(["test" : "test"], onChannel: nil, storeInHistory: true) { (status) -> Void in
            
            XCTAssertNotNil(status)
            XCTAssertTrue(status.category == .PNBadRequestCategory)
            XCTAssertTrue(status.operation == .PublishOperation)
            XCTAssertEqual(status.statusCode, 400)
            XCTAssertTrue(status.error)
            XCTAssertNil(status.data)
        }
    }
    
    func testPublishToNilChannelWithoutHistory() {
        self.performVerifiedPublish(["test" : "test"], onChannel: nil, storeInHistory: false) { (status) -> Void in
            
            XCTAssertNotNil(status)
            XCTAssertTrue(status.category == .PNBadRequestCategory)
            XCTAssertTrue(status.operation == .PublishOperation)
            XCTAssertEqual(status.statusCode, 400)
            XCTAssertTrue(status.error)
            XCTAssertNil(status.data)
        }
    }
}

extension PNPublishWithHistoryTests {
    func performVerifiedPublish(message: AnyObject?, onChannel: String?,
        storeInHistory: Bool, withAssertions: PNPublishCompletionBlock) {
            
            let networkExpectation = self.expectationWithDescription("network")
            
            self.client.publish(message, toChannel: onChannel, storeInHistory: storeInHistory) { (status) -> Void in
                withAssertions(status)
                networkExpectation.fulfill()
            }
            waitForExpectationsWithTimeout(5, handler: { (error) -> Void in
                XCTAssertNil(error, "Encountered error with publish call")
            })
    }
}

